# Assignment 3 - From Validated Files to Business Data Products

**Continuation of Assignment 1 (Olist Pandas Data Exploration, Processing & Validation)**

This notebook builds a small, trusted data system inside PostgreSQL (`olist_warehouse`):

```
Olist CSV files -> RAW -> CLEAN -> STAGING -> CURATED -> business answers
```

| Schema | Question it answers | What lives here |
|---|---|---|
| `raw` | What did we receive? | Source-shaped evidence (all TEXT, nothing interpreted) |
| `clean` | What trustworthy business entities do we have? | Typed, keyed, reusable entities |
| `staging` | What intermediate work is needed to build a data product? | Step-by-step transformation tables |
| `curated` | What final data product are we providing? | One table per business requirement |

**Tool choice rule used throughout:** Pandas for reading / inspecting / preparing files; SQL for schemas, table contracts, joins, staging and aggregation; either for evidence checks.

---
# 01. BUSINESS REQUIREMENTS

Written in plain English *before* any SQL. The same content is kept in `docs/business_requirements.md`.

## Business Problem 1 - Finance

| Field | Definition |
|---|---|
| **Business user** | Finance |
| **Business question** | How is item sales value changing across product categories over time? |
| **Metric - Item Sales Value** | `SUM(order_items.price)`. Item price only - **freight is excluded** because freight is a delivery charge, not product sales. |
| **Metric - Distinct Orders** | `COUNT(DISTINCT order_id)` inside each month + category. |
| **Metric - Items Sold** (extra, justified) | `COUNT(*)` of order items. Needed because one order can hold many items, so orders alone hide volume. |
| **Time definition** | Month of `orders.order_purchase_timestamp` (the moment the sale happened; it is never NULL). |
| **Included records** | Order items whose order status is **not** `canceled` or `unavailable`. |
| **Excluded records** | Items of `canceled` / `unavailable` orders (the sale did not happen). |
| **Unknown / missing category** | Product category is NULL for some products. In RAW/CLEAN it stays NULL. In the **staging** step it is explicitly labelled `unknown` so the sales value is not silently lost. |
| **Expected final grain** | One row = one product category in one purchase month. |
| **Assumptions** | (1) `price` is per item row, so quantity is already 1 per row. (2) An order with items from 2 categories counts once in **each** of those categories, so `distinct_orders` must **not** be summed across categories. (3) Statuses `created`, `approved`, `invoiced`, `processing`, `shipped`, `delivered` are all treated as real sales. (4) Category names stay in Portuguese (no translation file is part of this assignment). |

## Business Problem 2 - Operations

| Field | Definition |
|---|---|
| **Business user** | Operations |
| **Business question** | How is delivery performance changing, and where are late deliveries occurring? |
| **What is an order** | One row in `clean.orders`. The measure is based on **orders**, not order items (an order may have many item rows, which would over-count). |
| **Delivered** | `order_status = 'delivered'` **and** `order_delivered_customer_date` is not NULL. |
| **Late** | A delivered order where `order_delivered_customer_date` (as a date) is **after** `order_estimated_delivery_date` (as a date). |
| **Boundary rule** | Same day as estimate = **on time**. Dates are compared as dates because the estimate has no time part (midnight), so comparing timestamps would wrongly call a same-day afternoon delivery "late". |
| **Missing delivery date** | Delivered orders with a missing delivery date are classified `unknown` and are **not** counted as delivered or late (no evidence). |
| **Cancelled orders** | `canceled` and `unavailable` are classified `not_fulfilled`. They are never "late". |
| **Not yet delivered** | `created`, `approved`, `invoiced`, `processing`, `shipped` are `not_delivered` (still in progress). |
| **Reporting period** | Month of `order_purchase_timestamp` (every order has one; delivery month does not exist for undelivered orders). |
| **Location** | `customer_state` (27 values - readable; city would be too fragmented). |
| **Late Delivery Rate** | `late_orders / delivered_orders` (denominator = orders with a trustworthy delivery outcome). |
| **Expected final grain** | One row = one customer state in one purchase month. |

---
# 02. INSPECT EXISTING DATA

Load the Assignment 1 outputs, confirm the files exist and check columns before touching the database.

In [3]:
import sys
print(sys.executable)

c:\Users\PAVANI\OneDrive\Desktop\OneDrive\Desktop\olist-pandas-assignment\olist-pandas-assignment\.venv\Scripts\python.exe


In [6]:
import sys
!{sys.executable} -m pip install python-dotenv sqlalchemy psycopg[binary] ipykernel

   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--
   ---- ----------------------------------- 0.3/2.4 MB ? eta -:--:--
   ------------ --------------------------- 0.8/2.4 MB 1.6 MB/s eta 0:00:02
   ----------------- ---------------------- 1.0/2.4 MB 1.6 MB/s eta 0:00:01
   ------------------------- -------------- 1.6/2.4 MB 1.7 MB/s eta 0:00:01
   ------------------------------ --------- 1.8/2.4 MB 1.8 MB/s eta 0:00:01
   -------------------------------------- - 2.4/2.4 MB 1.7 MB/s eta 0:00:01
   ---------------------------------------- 2.4/2.4 MB 1.7 MB/s  0:00:01
   ---------------------------------------- 0.0/3.7 MB ? eta -:--:--
   -- ------------------------------------- 0.3/3.7 MB ? eta -:--:--
   ----- ---------------------------------- 0.5/3.7 MB 989.2 kB/s eta 0:00:04
   -------- ------------------------------- 0.8/3.7 MB 1.1 MB/s eta 0:00:03
   -------------- ------------------------- 1.3/

In [7]:
import csv
import io
import os
from pathlib import Path

import pandas as pd
from dotenv import load_dotenv
from sqlalchemy import create_engine, text

# Project root works whether Jupyter is started from the root or from /notebooks
ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
RAW_DIR = ROOT / "data" / "raw"            # original source files (evidence)
PROC_DIR = ROOT / "data" / "processed"     # Assignment 1 validated output
DOCS_DIR = ROOT / "docs"

FILES = {
    "customers":   "olist_customers_dataset.csv",
    "orders":      "olist_orders_dataset.csv",
    "order_items": "olist_order_items_dataset.csv",
    "products":    "olist_products_dataset.csv",
    "reviews":     "olist_order_reviews_dataset.csv",
}

pd.set_option("display.width", 200)
pd.set_option("display.max_columns", 30)
print("Project root:", ROOT)

Project root: c:\Users\PAVANI\OneDrive\Desktop\OneDrive\Desktop\olist-pandas-assignment\olist-pandas-assignment


In [ ]:
rows = []
for name, fname in FILES.items():
    raw_ok, proc_ok = (RAW_DIR / fname).exists(), (PROC_DIR / fname).exists()
    rows.append({"entity": name, "file": fname, "in_raw": raw_ok, "in_processed": proc_ok})
    assert raw_ok and proc_ok, f"Missing file for {name}"
pd.DataFrame(rows)

,entity,file,in_raw,in_processed
0,customers,olist_customers_dataset.csv,True,True
1,orders,olist_orders_dataset.csv,True,True
2,order_items,olist_order_items_dataset.csv,True,True
3,products,olist_products_dataset.csv,True,True
4,reviews,olist_order_reviews_dataset.csv,True,True


In [ ]:
# Columns and dtypes of the Assignment 1 processed files
processed_preview = {}
for name, fname in FILES.items():
    df = pd.read_csv(PROC_DIR / fname, nrows=1000)
    processed_preview[name] = df
    print(f"--- {name}: {list(df.columns)}")

--- customers: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
--- orders: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
--- order_items: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
--- products: ['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
--- reviews: ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']


---
# 03. CONNECT TO POSTGRESQL & CREATE SCHEMAS

Credentials are read from a `.env` file (git-ignored) - never typed into the notebook.
Create `.env` from `.env.example`:

```
DATABASE_URL=postgresql://<username>:<password>@localhost:<port>/olist_warehouse
```

The database must exist first (run once in pgAdmin / psql): `CREATE DATABASE olist_warehouse;`

### Pause and Observe - why is a real password inside the notebook dangerous?

If `connection = "postgresql://user:realpassword@..."` is written in a notebook and committed to GitHub:

1. **Git keeps history forever.** Even if I delete the line later, the password stays in old commits and anyone with repo access can read it.
2. **Repositories get shared or become public** (teammates, forks, portfolio, mistakes). Bots scan GitHub for leaked credentials within minutes.
3. **The same password is often reused**, so one leak can expose other systems.
4. **Notebooks also save outputs** - an error message or printed connection string can leak the secret too.
5. Environment variables / `.env` keep secrets **outside** the code, so each person (or server) uses its own credentials and a leaked password can be rotated without changing code.

In [ ]:
load_dotenv(ROOT / ".env")
DATABASE_URL = os.getenv("DATABASE_URL")
assert DATABASE_URL, "DATABASE_URL not found. Create a .env file (see .env.example)."

# SQLAlchemy needs to be told which driver to use (psycopg 3)
if DATABASE_URL.startswith("postgresql://"):
    DATABASE_URL = DATABASE_URL.replace("postgresql://", "postgresql+psycopg://", 1)

engine = create_engine(DATABASE_URL)


# ---------- small helpers used in every section ----------
def run_sql(sql):
    """Run one or more SQL statements (DDL / DML) in a single transaction."""
    statements = [s.strip() for s in sql.split(";") if s.strip()]
    with engine.begin() as conn:
        for stmt in statements:
            conn.execute(text(stmt))


def query_df(sql):
    """Run a SELECT and return a DataFrame."""
    with engine.connect() as conn:
        return pd.read_sql(text(sql), conn)


def scalar(sql):
    """Run a SELECT that returns one value."""
    with engine.connect() as conn:
        return conn.execute(text(sql)).scalar()


def copy_df(df, table, columns):
    """Bulk-load a DataFrame with PostgreSQL COPY (one transaction, much faster than row inserts)."""
    buf = io.StringIO()
    df[columns].to_csv(buf, index=False, header=False, na_rep="")   # missing -> empty -> NULL
    col_list = ", ".join(columns)
    raw_conn = engine.raw_connection()
    try:
        with raw_conn.cursor() as cur:
            with cur.copy(f"COPY {table} ({col_list}) FROM STDIN WITH (FORMAT csv, NULL '')") as cp:
                cp.write(buf.getvalue())
        raw_conn.commit()
    except Exception:
        raw_conn.rollback()
        raise
    finally:
        raw_conn.close()


# ---------- validation recorder (feeds docs/validation_evidence.md) ----------
CHECKS = []


def check(stage, description, passed, detail=""):
    CHECKS.append({"stage": stage, "check": description,
                   "result": "PASS" if passed else "FAIL", "detail": str(detail)})
    print(("PASS" if passed else "FAIL"), "|", stage, "|", description, "|", detail)
    assert passed, f"Validation failed: {stage} - {description} ({detail})"


def null_count(table, column):
    return scalar(f"SELECT COUNT(*) FROM {table} WHERE {column} IS NULL")


def dup_count(table, columns):
    cols = ", ".join(columns)
    return scalar(f"SELECT COUNT(*) FROM (SELECT {cols} FROM {table} GROUP BY {cols} HAVING COUNT(*) > 1) d")


print(query_df("SELECT current_database() AS connected_database").to_string(index=False))

connected_database
   olist_warehouse


In [ ]:
sql_schemas = """
CREATE SCHEMA IF NOT EXISTS raw;
CREATE SCHEMA IF NOT EXISTS clean;
CREATE SCHEMA IF NOT EXISTS staging;
CREATE SCHEMA IF NOT EXISTS curated
"""
run_sql(sql_schemas)
query_df("""
    SELECT schema_name FROM information_schema.schemata
    WHERE schema_name IN ('raw','clean','staging','curated') ORDER BY schema_name
""")

,schema_name
0,clean
1,curated
2,raw
3,staging


---
# 04. LOAD RAW & BUILD CLEAN

## 04-A RAW - "What did we receive?"

Design rules for RAW:

* Source = the **original files in `data/raw/`** (the evidence), not the processed copies.
* Every source column is `TEXT`. No type conversion, no replacing missing values, no filtering.
* Two traceability columns are added by the database: `_source_file` and `_loaded_at`.
* Reloadable: each run drops and recreates the table inside one transaction, so loading twice cannot create duplicates.

**Prediction:** each RAW table row count = number of data records in its CSV (header excluded). Reviews will **not** equal `wc -l`, because review comments contain line breaks inside quotes - so I count records with Python's `csv` module, not by counting lines.

In [ ]:
def count_csv_records(path):
    """Independent record count (csv parser, header excluded). Handles line breaks inside quoted text."""
    with open(path, newline="", encoding="utf-8") as f:
        return sum(1 for _ in csv.reader(f)) - 1


raw_frames = {}
for name, fname in FILES.items():
    # dtype=str -> Pandas does NOT guess types; we keep exactly what the file contains
    df = pd.read_csv(RAW_DIR / fname, dtype=str)
    raw_frames[name] = df
    print(f"{name:12s} rows={len(df):>8,}  cols={df.shape[1]}  missing cells={int(df.isna().sum().sum()):,}")

customers    rows=  99,441  cols=5  missing cells=0


orders       rows=  99,441  cols=8  missing cells=4,908


order_items  rows= 112,650  cols=7  missing cells=0
products     rows=  32,951  cols=9  missing cells=2,448


reviews      rows=  99,224  cols=7  missing cells=145,903


In [ ]:
# RAW DDL is generated from the file header so the table is exactly source-shaped
raw_ddl = []
for name, fname in FILES.items():
    cols = ",\n    ".join(f'"{c}" TEXT' for c in raw_frames[name].columns)
    raw_ddl.append(
        f"DROP TABLE IF EXISTS raw.{name} CASCADE;\n"
        f"CREATE TABLE raw.{name} (\n    {cols},\n"
        f"    _source_file TEXT DEFAULT '{fname}',\n"
        f"    _loaded_at   TIMESTAMPTZ DEFAULT now()\n)"
    )
run_sql(";\n".join(raw_ddl))
print(raw_ddl[0])

DROP TABLE IF EXISTS raw.customers CASCADE;
CREATE TABLE raw.customers (
    "customer_id" TEXT,
    "customer_unique_id" TEXT,
    "customer_zip_code_prefix" TEXT,
    "customer_city" TEXT,
    "customer_state" TEXT,
    _source_file TEXT DEFAULT 'olist_customers_dataset.csv',
    _loaded_at   TIMESTAMPTZ DEFAULT now()
)


In [ ]:
# Load RAW with COPY. Each table: clear + load in one go, so re-running never duplicates rows.
for name, df in raw_frames.items():
    run_sql(f"TRUNCATE raw.{name}")
    copy_df(df, f"raw.{name}", list(df.columns))
    print(f"loaded raw.{name}")

loaded raw.customers


loaded raw.orders


loaded raw.order_items
loaded raw.products


loaded raw.reviews


### RAW validation - source file row count vs database row count
If they differ: stop and investigate (header counted? loaded twice? parse error? transaction not committed? wrong file?).

In [ ]:
for name, fname in FILES.items():
    file_rows = count_csv_records(RAW_DIR / fname)
    db_rows = scalar(f"SELECT COUNT(*) FROM raw.{name}")
    check("RAW", f"raw.{name}: source file rows = database rows", file_rows == db_rows,
          f"file={file_rows:,} db={db_rows:,}")

# RAW must keep missing values as NULL and NOT rewrite them
check("RAW", "raw.products keeps missing category as NULL (not 'Unknown')",
      scalar("SELECT COUNT(*) FROM raw.products WHERE product_category_name IS NULL") == 610
      and scalar("SELECT COUNT(*) FROM raw.products WHERE product_category_name ILIKE 'unknown'") == 0,
      "610 NULL categories, 0 'unknown'")
check("RAW", "raw.orders keeps all 8 order statuses untouched",
      scalar("SELECT COUNT(DISTINCT order_status) FROM raw.orders") == 8, "8 statuses")
query_df("SELECT * FROM raw.orders ORDER BY order_id LIMIT 3")

PASS | RAW | raw.customers: source file rows = database rows | file=99,441 db=99,441


PASS | RAW | raw.orders: source file rows = database rows | file=99,441 db=99,441


PASS | RAW | raw.order_items: source file rows = database rows | file=112,650 db=112,650


PASS | RAW | raw.products: source file rows = database rows | file=32,951 db=32,951


PASS | RAW | raw.reviews: source file rows = database rows | file=99,224 db=99,224


PASS | RAW | raw.products keeps missing category as NULL (not 'Unknown') | 610 NULL categories, 0 'unknown'
PASS | RAW | raw.orders keeps all 8 order statuses untouched | 8 statuses


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,_source_file,_loaded_at
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,delivered,2017-09-13 08:59:02,2017-09-13 09:45:35,2017-09-19 18:34:16,2017-09-20 23:43:48,2017-09-29 00:00:00,olist_orders_dataset.csv,2026-10-02 10:15:46.666692+00:00
1,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,delivered,2017-04-26 10:53:06,2017-04-26 11:05:13,2017-05-04 14:35:00,2017-05-12 16:04:24,2017-05-15 00:00:00,olist_orders_dataset.csv,2026-10-02 10:15:46.666692+00:00
2,000229ec398224ef6ca0657da4fc703e,6489ae5e4333f3693df5ad4372dab6d3,delivered,2018-01-14 14:33:31,2018-01-14 14:48:30,2018-01-16 12:36:48,2018-01-22 13:19:16,2018-02-05 00:00:00,olist_orders_dataset.csv,2026-10-02 10:15:46.666692+00:00


## 04-B CLEAN - "What trustworthy business entities do we have?"

CLEAN is built from the **Assignment 1 processed files** (validated) and given real data types, keys and relationships.
It is **not** shaped around Finance or Operations - any team can reuse it.

### Design before loading

| Table | Purpose | Grain (one row =) | Primary key | Foreign keys |
|---|---|---|---|---|
| `clean.customers` | Customer records attached to orders, with location | one customer record (`customer_id`) | `customer_id` | - |
| `clean.products` | Product catalogue + category + physical attributes | one product | `product_id` | - |
| `clean.orders` | Order header: status and lifecycle dates | one order | `order_id` | `customer_id -> clean.customers` |
| `clean.order_items` | Items sold inside orders (the sales event) | one item line of one order | `(order_id, order_item_id)` | `order_id -> clean.orders`, `product_id -> clean.products` |
| `clean.reviews` | Customer review of an order | one review for one order | `(review_id, order_id)` | `order_id -> clean.orders` |

**Real-world rules behind the constraints**

* `customer_id` is unique per order-customer record (99,441), while `customer_unique_id` repeats (96,096) because one real person can place several orders - so only `customer_id` is the key and both are kept.
* An order item is identified by order + its line number, so the key is composite.
* `price > 0` and `freight_value >= 0`: a product cannot be sold for 0 or less (Assignment 1 found min price 0.85), but free shipping is real (383 rows with 0 freight).
* `review_score` between 1 and 5 is the rating scale.
* `review_id` alone is **not** unique (98,410 unique IDs vs 99,224 rows, found in Assignment 1) - the same review ID appears on several orders, so the key is `(review_id, order_id)`.
* Missing category, missing delivery dates and missing review comments are **allowed** (nullable) - Assignment 1 decided to keep them; they carry business meaning.
* Dates are `TIMESTAMP` (the source has no time zone). Money is `NUMERIC` (exact), never float.

**Load order (dependency-aware):** `customers -> products -> orders -> order_items -> reviews`.
Parents first: `orders` needs `customers`; `order_items` needs `orders` and `products`; `reviews` needs `orders`.

In [ ]:
sql_clean_ddl = """
DROP TABLE IF EXISTS clean.reviews CASCADE;
DROP TABLE IF EXISTS clean.order_items CASCADE;
DROP TABLE IF EXISTS clean.orders CASCADE;
DROP TABLE IF EXISTS clean.products CASCADE;
DROP TABLE IF EXISTS clean.customers CASCADE;

CREATE TABLE clean.customers (
    customer_id              TEXT PRIMARY KEY,
    customer_unique_id       TEXT NOT NULL,
    customer_zip_code_prefix INTEGER,
    customer_city            TEXT,
    customer_state           CHAR(2) NOT NULL
);

CREATE TABLE clean.products (
    product_id                 TEXT PRIMARY KEY,
    product_category_name      TEXT,
    product_name_lenght        INTEGER,
    product_description_lenght INTEGER,
    product_photos_qty         INTEGER,
    product_weight_g           INTEGER,
    product_length_cm          INTEGER,
    product_height_cm          INTEGER,
    product_width_cm           INTEGER
);

CREATE TABLE clean.orders (
    order_id                      TEXT PRIMARY KEY,
    customer_id                   TEXT NOT NULL REFERENCES clean.customers (customer_id),
    order_status                  TEXT NOT NULL,
    order_purchase_timestamp      TIMESTAMP NOT NULL,
    order_approved_at             TIMESTAMP,
    order_delivered_carrier_date  TIMESTAMP,
    order_delivered_customer_date TIMESTAMP,
    order_estimated_delivery_date TIMESTAMP NOT NULL
);

CREATE TABLE clean.order_items (
    order_id            TEXT NOT NULL REFERENCES clean.orders (order_id),
    order_item_id       INTEGER NOT NULL CHECK (order_item_id >= 1),
    product_id          TEXT NOT NULL REFERENCES clean.products (product_id),
    seller_id           TEXT NOT NULL,
    shipping_limit_date TIMESTAMP,
    price               NUMERIC(10,2) NOT NULL CHECK (price > 0),
    freight_value       NUMERIC(10,2) NOT NULL CHECK (freight_value >= 0),
    PRIMARY KEY (order_id, order_item_id)
);

CREATE TABLE clean.reviews (
    review_id               TEXT NOT NULL,
    order_id                TEXT NOT NULL REFERENCES clean.orders (order_id),
    review_score            SMALLINT NOT NULL CHECK (review_score BETWEEN 1 AND 5),
    review_comment_title    TEXT,
    review_comment_message  TEXT,
    review_creation_date    TIMESTAMP,
    review_answer_timestamp TIMESTAMP,
    PRIMARY KEY (review_id, order_id)
)
"""
run_sql(sql_clean_ddl)

### Prepare dataframes with Pandas (types only - no business changes)
Assignment 1 wrote the processed files as CSV, so types were lost. I restore them here: dates -> datetime, product measurements -> nullable integer (`Int64`, because they contain missing values and CSV stored them as `40.0`).

In [ ]:
clean_frames = {}

clean_frames["customers"] = pd.read_csv(PROC_DIR / FILES["customers"])

products = pd.read_csv(PROC_DIR / FILES["products"])
int_cols = ["product_name_lenght", "product_description_lenght", "product_photos_qty", "product_weight_g",
            "product_length_cm", "product_height_cm", "product_width_cm"]
for c in int_cols:
    products[c] = products[c].astype("Int64")
clean_frames["products"] = products

clean_frames["orders"] = pd.read_csv(
    PROC_DIR / FILES["orders"],
    parse_dates=["order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date",
                 "order_delivered_customer_date", "order_estimated_delivery_date"])

clean_frames["order_items"] = pd.read_csv(PROC_DIR / FILES["order_items"], parse_dates=["shipping_limit_date"])

clean_frames["reviews"] = pd.read_csv(
    PROC_DIR / FILES["reviews"], parse_dates=["review_creation_date", "review_answer_timestamp"])

for name, df in clean_frames.items():
    print(f"{name:12s} rows={len(df):>8,}")

customers    rows=  99,441
products     rows=  32,951
orders       rows=  99,441
order_items  rows= 112,650
reviews      rows=  99,224


In [ ]:
# Load in dependency order (parents before children)
LOAD_ORDER = ["customers", "products", "orders", "order_items", "reviews"]
for name in LOAD_ORDER:
    df = clean_frames[name]
    copy_df(df, f"clean.{name}", list(df.columns))
    n_rows = scalar(f"SELECT COUNT(*) FROM clean.{name}")
    print(f"loaded clean.{name}: {n_rows:,} rows")

loaded clean.customers: 99,441 rows
loaded clean.products: 32,951 rows


loaded clean.orders: 99,441 rows


loaded clean.order_items: 112,650 rows


loaded clean.reviews: 99,224 rows


### CLEAN validation
Identity (keys non-null + unique), relationships (no orphans), types, grain, and reconciliation against the validated processed files.

In [ ]:
# --- Identity: primary keys non-null and unique ---
PKS = {
    "customers":   ["customer_id"],
    "products":    ["product_id"],
    "orders":      ["order_id"],
    "order_items": ["order_id", "order_item_id"],
    "reviews":     ["review_id", "order_id"],
}
for table, cols in PKS.items():
    nulls = sum(null_count(f"clean.{table}", c) for c in cols)
    dups = dup_count(f"clean.{table}", cols)
    check("CLEAN", f"clean.{table}: PK {tuple(cols)} has no NULLs and no duplicates (grain)", nulls == 0 and dups == 0,
          f"nulls={nulls} duplicate_keys={dups}")

PASS | CLEAN | clean.customers: PK ('customer_id',) has no NULLs and no duplicates (grain) | nulls=0 duplicate_keys=0
PASS | CLEAN | clean.products: PK ('product_id',) has no NULLs and no duplicates (grain) | nulls=0 duplicate_keys=0
PASS | CLEAN | clean.orders: PK ('order_id',) has no NULLs and no duplicates (grain) | nulls=0 duplicate_keys=0


PASS | CLEAN | clean.order_items: PK ('order_id', 'order_item_id') has no NULLs and no duplicates (grain) | nulls=0 duplicate_keys=0


PASS | CLEAN | clean.reviews: PK ('review_id', 'order_id') has no NULLs and no duplicates (grain) | nulls=0 duplicate_keys=0


In [ ]:
# --- Relationships: every child key finds its parent ---
RELS = [
    ("orders.customer_id -> customers",   "clean.orders o",      "clean.customers p", "o.customer_id = p.customer_id", "p.customer_id"),
    ("order_items.order_id -> orders",    "clean.order_items o", "clean.orders p",    "o.order_id = p.order_id",       "p.order_id"),
    ("order_items.product_id -> products", "clean.order_items o", "clean.products p", "o.product_id = p.product_id",   "p.product_id"),
    ("reviews.order_id -> orders",        "clean.reviews o",     "clean.orders p",    "o.order_id = p.order_id",       "p.order_id"),
]
for label, child, parent, cond, pk in RELS:
    orphans = scalar(f"SELECT COUNT(*) FROM {child} LEFT JOIN {parent} ON {cond} WHERE {pk} IS NULL")
    check("CLEAN", f"FK {label}: no orphan rows", orphans == 0, f"orphans={orphans}")

# Orders without items are real (cancelled / unavailable etc.) - documented, not an error
no_items = scalar("SELECT COUNT(*) FROM clean.orders o LEFT JOIN clean.order_items i ON i.order_id = o.order_id WHERE i.order_id IS NULL")
print(f"Info: {no_items:,} orders have no order_items (expected: mostly unavailable / canceled orders).")
query_df("""
    SELECT o.order_status, COUNT(*) AS orders_without_items
    FROM clean.orders o LEFT JOIN clean.order_items i ON i.order_id = o.order_id
    WHERE i.order_id IS NULL GROUP BY o.order_status ORDER BY 2 DESC
""")

PASS | CLEAN | FK orders.customer_id -> customers: no orphan rows | orphans=0


PASS | CLEAN | FK order_items.order_id -> orders: no orphan rows | orphans=0
PASS | CLEAN | FK order_items.product_id -> products: no orphan rows | orphans=0


PASS | CLEAN | FK reviews.order_id -> orders: no orphan rows | orphans=0


Info: 775 orders have no order_items (expected: mostly unavailable / canceled orders).


,order_status,orders_without_items
0,unavailable,603
1,canceled,164
2,created,5
3,invoiced,2
4,shipped,1


In [ ]:
# --- Types: dates are dates, money is numeric ---
types = query_df("""
    SELECT table_name, column_name, data_type
    FROM information_schema.columns
    WHERE table_schema = 'clean' ORDER BY table_name, ordinal_position
""")
expected_types = {
    ("orders", "order_purchase_timestamp"): "timestamp without time zone",
    ("orders", "order_delivered_customer_date"): "timestamp without time zone",
    ("orders", "order_estimated_delivery_date"): "timestamp without time zone",
    ("order_items", "price"): "numeric",
    ("order_items", "freight_value"): "numeric",
    ("order_items", "shipping_limit_date"): "timestamp without time zone",
    ("reviews", "review_score"): "smallint",
    ("reviews", "review_creation_date"): "timestamp without time zone",
    ("products", "product_weight_g"): "integer",
}
actual = {(r.table_name, r.column_name): r.data_type for r in types.itertuples()}
bad = {k: (v, actual.get(k)) for k, v in expected_types.items() if actual.get(k) != v}
check("CLEAN", "key columns have the intended data types", not bad, f"mismatches={bad}")
types[types.table_name.isin(["orders", "order_items"])]

PASS | CLEAN | key columns have the intended data types | mismatches={}


,table_name,column_name,data_type
5,order_items,order_id,text
6,order_items,order_item_id,integer
7,order_items,product_id,text
8,order_items,seller_id,text
9,order_items,shipping_limit_date,timestamp without time zone
10,order_items,price,numeric
11,order_items,freight_value,numeric
12,orders,order_id,text
13,orders,customer_id,text
14,orders,order_status,text


In [ ]:
# --- Reconciliation: CLEAN vs validated processed files (row counts + money total) ---
for name in LOAD_ORDER:
    file_rows = len(pd.read_csv(PROC_DIR / FILES[name]))
    db_rows = scalar(f"SELECT COUNT(*) FROM clean.{name}")
    check("CLEAN", f"clean.{name}: row count = processed file", file_rows == db_rows, f"file={file_rows:,} db={db_rows:,}")

price_file = round(pd.read_csv(PROC_DIR / FILES["order_items"])["price"].sum(), 2)
price_db = float(scalar("SELECT SUM(price) FROM clean.order_items"))
check("CLEAN", "SUM(price): clean.order_items = processed file", abs(price_file - price_db) < 0.01,
      f"file={price_file:,.2f} db={price_db:,.2f}")

# --- Reconciliation: CLEAN vs RAW (nothing lost between layers) ---
for name in LOAD_ORDER:
    check("CLEAN", f"clean.{name} row count = raw.{name} row count",
          scalar(f"SELECT COUNT(*) FROM clean.{name}") == scalar(f"SELECT COUNT(*) FROM raw.{name}"))

PASS | CLEAN | clean.customers: row count = processed file | file=99,441 db=99,441


PASS | CLEAN | clean.products: row count = processed file | file=32,951 db=32,951


PASS | CLEAN | clean.orders: row count = processed file | file=99,441 db=99,441


PASS | CLEAN | clean.order_items: row count = processed file | file=112,650 db=112,650


PASS | CLEAN | clean.reviews: row count = processed file | file=99,224 db=99,224


PASS | CLEAN | SUM(price): clean.order_items = processed file | file=13,591,643.70 db=13,591,643.70
PASS | CLEAN | clean.customers row count = raw.customers row count | 
PASS | CLEAN | clean.products row count = raw.products row count | 
PASS | CLEAN | clean.orders row count = raw.orders row count | 
PASS | CLEAN | clean.order_items row count = raw.order_items row count | 
PASS | CLEAN | clean.reviews row count = raw.reviews row count | 


---
# 05. FINANCE PIPELINE

**Requirement -> Grain -> CLEAN sources -> STAGING 1 -> STAGING 2 -> CURATED -> Business query** (validate at every step)

## Step 1 - Understand
Finance wants to see **item sales value** by **product category** by **month**. Definitions are in section 01.

## Step 2 - Grain thinking exercise (answers *before* writing SQL)

1. **Grain of the curated table?** One row = one product category in one purchase month.
2. **What does `item_sales_value` mean?** Sum of `order_items.price` for eligible items. Freight excluded.
3. **Which timestamp decides the month?** `order_purchase_timestamp` from `clean.orders`.
4. **Which records are excluded?** Items belonging to `canceled` or `unavailable` orders.
5. **Can one order contribute multiple rows to the detailed staging table?** Yes - an order with 3 items gives 3 item rows.
6. **How to avoid counting the order many times?** Use `COUNT(DISTINCT order_id)` at aggregation; never `COUNT(*)` for orders.
7. **Products without a category?** Kept (their sales still count). Labelled `unknown` in staging 2 - an explicit, documented decision, not hidden.
8. **How to prove curated matches CLEAN?** Re-calculate the same totals straight from `clean.order_items JOIN clean.orders` (no staging) and compare totals, per-month values and per-category values.

## Step 3 - Source entities
Only three CLEAN tables contribute: `clean.order_items` (price = the sales event), `clean.orders` (status + purchase timestamp), `clean.products` (category).

## Step 4 - Relationships
`order_items.order_id -> orders.order_id` (many items to one order) and `order_items.product_id -> products.product_id` (many items to one product). Both joins are **many-to-one from the item side**.

## STAGING 1 - `staging.finance_items_enriched`
* **One row =** one order item (`order_id`, `order_item_id`).
* **Why it exists:** gathers everything needed to reason about a sale in one place (price + status + purchase time + raw category). No business filtering yet, so it can be compared 1:1 with `clean.order_items`.
* **Prediction:** exactly the same number of rows as `clean.order_items`. I start from the items (the *many* side) and join to parents (`orders`, `products`) that each match once, so rows cannot multiply. I use `LEFT JOIN` for products so an item is never dropped if a product is missing.

In [ ]:
sql_fin_stg1 = """
DROP TABLE IF EXISTS staging.finance_items_enriched;

CREATE TABLE staging.finance_items_enriched AS
SELECT
    oi.order_id,
    oi.order_item_id,
    oi.product_id,
    o.order_status,
    o.order_purchase_timestamp,
    oi.price,
    oi.freight_value,
    p.product_category_name
FROM clean.order_items oi
JOIN clean.orders      o ON o.order_id   = oi.order_id
LEFT JOIN clean.products p ON p.product_id = oi.product_id;

ALTER TABLE staging.finance_items_enriched ADD PRIMARY KEY (order_id, order_item_id)
"""
run_sql(sql_fin_stg1)

In [ ]:
# STAGING 1 validation: row count, grain, nulls, join behaviour, totals, sample
expected = scalar("SELECT COUNT(*) FROM clean.order_items")
actual = scalar("SELECT COUNT(*) FROM staging.finance_items_enriched")
check("FIN-STG1", "row count = clean.order_items (joins did not multiply or drop rows)", expected == actual,
      f"expected={expected:,} actual={actual:,}")
check("FIN-STG1", "declared grain unique (order_id, order_item_id)",
      dup_count("staging.finance_items_enriched", ["order_id", "order_item_id"]) == 0)
for col in ["order_id", "order_status", "order_purchase_timestamp", "price"]:
    check("FIN-STG1", f"no NULLs in {col}", null_count("staging.finance_items_enriched", col) == 0)

null_cat = null_count("staging.finance_items_enriched", "product_category_name")
print(f"Info: {null_cat:,} item rows have a NULL category (products without a category) - handled in STAGING 2.")

check("FIN-STG1", "SUM(price) = clean.order_items",
      scalar("SELECT SUM(price) FROM staging.finance_items_enriched") == scalar("SELECT SUM(price) FROM clean.order_items"))
query_df("SELECT * FROM staging.finance_items_enriched ORDER BY order_id, order_item_id LIMIT 5")

PASS | FIN-STG1 | row count = clean.order_items (joins did not multiply or drop rows) | expected=112,650 actual=112,650
PASS | FIN-STG1 | declared grain unique (order_id, order_item_id) | 
PASS | FIN-STG1 | no NULLs in order_id | 
PASS | FIN-STG1 | no NULLs in order_status | 
PASS | FIN-STG1 | no NULLs in order_purchase_timestamp | 
PASS | FIN-STG1 | no NULLs in price | 
Info: 1,603 item rows have a NULL category (products without a category) - handled in STAGING 2.
PASS | FIN-STG1 | SUM(price) = clean.order_items | 


,order_id,order_item_id,product_id,order_status,order_purchase_timestamp,price,freight_value,product_category_name
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,delivered,2017-09-13 08:59:02,58.90,13.29,cool_stuff
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,delivered,2017-04-26 10:53:06,239.90,19.93,pet_shop
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,delivered,2018-01-14 14:33:31,199.00,17.87,moveis_decoracao
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,delivered,2018-08-08 10:00:35,12.99,12.79,perfumaria
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,delivered,2017-02-04 13:57:51,199.90,18.14,ferramentas_jardim


## STAGING 2 - `staging.finance_items_reportable`
* **One row =** one *reportable* order item (still item level).
* **Why it exists:** this is where the Finance business rules are applied, separately from the joining work: (1) keep only eligible statuses, (2) derive `reporting_month`, (3) resolve the missing category to `unknown`, (4) expose the measure as `item_sales_value`.
* **What changed from staging 1:** fewer rows (cancelled/unavailable removed), new derived columns, freight and raw timestamps no longer needed.
* **Where aggregation belongs:** *not here*. Keeping item level here lets me inspect individual rows and keeps `COUNT(DISTINCT order_id)` possible. Aggregation happens once, in CURATED.
* **Prediction:** rows = staging 1 rows minus the items of `canceled`/`unavailable` orders. I compute that number independently from CLEAN to compare.

In [ ]:
sql_fin_stg2 = """
DROP TABLE IF EXISTS staging.finance_items_reportable;

CREATE TABLE staging.finance_items_reportable AS
SELECT
    order_id,
    order_item_id,
    order_status,
    CAST(date_trunc('month', order_purchase_timestamp) AS date)        AS reporting_month,
    COALESCE(NULLIF(TRIM(product_category_name), ''), 'unknown')         AS product_category,
    price                                                                AS item_sales_value
FROM staging.finance_items_enriched
WHERE order_status NOT IN ('canceled', 'unavailable');

ALTER TABLE staging.finance_items_reportable ADD PRIMARY KEY (order_id, order_item_id)
"""
run_sql(sql_fin_stg2)

In [ ]:
# STAGING 2 validation
expected = scalar("""
    SELECT COUNT(*) FROM clean.order_items oi JOIN clean.orders o ON o.order_id = oi.order_id
    WHERE o.order_status NOT IN ('canceled', 'unavailable')
""")
actual = scalar("SELECT COUNT(*) FROM staging.finance_items_reportable")
stg1 = scalar("SELECT COUNT(*) FROM staging.finance_items_enriched")
check("FIN-STG2", "row count = items of eligible orders (computed independently from CLEAN)", expected == actual,
      f"expected={expected:,} actual={actual:,} (removed {stg1 - actual:,} items from canceled/unavailable orders)")
check("FIN-STG2", "declared grain unique (order_id, order_item_id)",
      dup_count("staging.finance_items_reportable", ["order_id", "order_item_id"]) == 0)
for col in ["reporting_month", "product_category", "item_sales_value"]:
    check("FIN-STG2", f"no NULLs in {col}", null_count("staging.finance_items_reportable", col) == 0)

unknown_expected = scalar("""
    SELECT COUNT(*) FROM clean.order_items oi
    JOIN clean.orders o ON o.order_id = oi.order_id
    JOIN clean.products p ON p.product_id = oi.product_id
    WHERE o.order_status NOT IN ('canceled','unavailable') AND p.product_category_name IS NULL
""")
unknown_actual = scalar("SELECT COUNT(*) FROM staging.finance_items_reportable WHERE product_category = 'unknown'")
check("FIN-STG2", "items with missing category are labelled 'unknown' (none lost)", unknown_expected == unknown_actual,
      f"unknown items={unknown_actual:,}")
query_df("SELECT * FROM staging.finance_items_reportable ORDER BY order_id, order_item_id LIMIT 5")

PASS | FIN-STG2 | row count = items of eligible orders (computed independently from CLEAN) | expected=112,101 actual=112,101 (removed 549 items from canceled/unavailable orders)
PASS | FIN-STG2 | declared grain unique (order_id, order_item_id) | 
PASS | FIN-STG2 | no NULLs in reporting_month | 
PASS | FIN-STG2 | no NULLs in product_category | 
PASS | FIN-STG2 | no NULLs in item_sales_value | 
PASS | FIN-STG2 | items with missing category are labelled 'unknown' (none lost) | unknown items=1,589


,order_id,order_item_id,order_status,reporting_month,product_category,item_sales_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,delivered,2017-09-01,cool_stuff,58.90
1,00018f77f2f0320c557190d7a144bdd3,1,delivered,2017-04-01,pet_shop,239.90
2,000229ec398224ef6ca0657da4fc703e,1,delivered,2018-01-01,moveis_decoracao,199.00
3,00024acbcdf0a6daa1e931b038114c75,1,delivered,2018-08-01,perfumaria,12.99
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,delivered,2017-02-01,ferramentas_jardim,199.90


## CURATED - `curated.finance_monthly_category_sales`
* **Purpose:** answer Finance's recurring question without needing to understand the relational model.
* **One row =** one product category in one purchase month.
* **Columns / metrics**
  * `reporting_month` - first day of the purchase month
  * `product_category` - category name (`unknown` when the product has no category)
  * `item_sales_value` - `SUM(price)`, freight excluded
  * `items_sold` - number of item lines
  * `distinct_orders` - number of different orders that contain at least one item of that category in that month (**do not sum across categories**)
* **Prediction:** row count = number of distinct (month, category) pairs in staging 2.

In [ ]:
sql_fin_curated = """
DROP TABLE IF EXISTS curated.finance_monthly_category_sales;

CREATE TABLE curated.finance_monthly_category_sales (
    reporting_month   DATE          NOT NULL,
    product_category  TEXT          NOT NULL,
    item_sales_value  NUMERIC(14,2) NOT NULL,
    items_sold        INTEGER       NOT NULL,
    distinct_orders   INTEGER       NOT NULL,
    PRIMARY KEY (reporting_month, product_category)
);

INSERT INTO curated.finance_monthly_category_sales
SELECT
    reporting_month,
    product_category,
    SUM(item_sales_value),
    COUNT(*),
    COUNT(DISTINCT order_id)
FROM staging.finance_items_reportable
GROUP BY reporting_month, product_category;

COMMENT ON TABLE curated.finance_monthly_category_sales IS
  'Finance product. Grain: one product category per purchase month. item_sales_value = SUM(price), freight excluded, canceled/unavailable orders excluded, missing category = unknown. distinct_orders must not be summed across categories.'
"""
run_sql(sql_fin_curated)

### Reconcile - can I prove CURATED agrees with CLEAN under the same business rules?
The comparison queries below go **directly to CLEAN** (no staging tables), so they test the whole chain.

In [ ]:
# 1) Grain / row count
expected_rows = scalar("SELECT COUNT(*) FROM (SELECT DISTINCT reporting_month, product_category FROM staging.finance_items_reportable) d")
actual_rows = scalar("SELECT COUNT(*) FROM curated.finance_monthly_category_sales")
check("FIN-CUR", "curated row count = distinct (month, category) pairs", expected_rows == actual_rows, f"rows={actual_rows:,}")
check("FIN-CUR", "no duplicate (reporting_month, product_category)",
      dup_count("curated.finance_monthly_category_sales", ["reporting_month", "product_category"]) == 0)

# 2) Total sales: curated vs CLEAN (independent query)
clean_total = scalar("""
    SELECT SUM(oi.price) FROM clean.order_items oi
    JOIN clean.orders o ON o.order_id = oi.order_id
    WHERE o.order_status NOT IN ('canceled', 'unavailable')
""")
cur_total = scalar("SELECT SUM(item_sales_value) FROM curated.finance_monthly_category_sales")
check("FIN-CUR", "total item_sales_value = CLEAN total (same rules)", clean_total == cur_total,
      f"clean={clean_total:,} curated={cur_total:,}")
check("FIN-CUR", "total items_sold = CLEAN eligible item count",
      scalar("SELECT SUM(items_sold) FROM curated.finance_monthly_category_sales") ==
      scalar("SELECT COUNT(*) FROM clean.order_items oi JOIN clean.orders o USING (order_id) WHERE o.order_status NOT IN ('canceled','unavailable')"))

# 3) Month x category level: rows that differ in either direction (must be 0)
diff_sql = """
WITH clean_calc AS (
    SELECT CAST(date_trunc('month', o.order_purchase_timestamp) AS date) AS m,
           COALESCE(p.product_category_name, 'unknown') AS c,
           SUM(oi.price) AS v, COUNT(*) AS n, COUNT(DISTINCT oi.order_id) AS d
    FROM clean.order_items oi
    JOIN clean.orders o   ON o.order_id = oi.order_id
    JOIN clean.products p ON p.product_id = oi.product_id
    WHERE o.order_status NOT IN ('canceled', 'unavailable')
    GROUP BY 1, 2
)
SELECT COUNT(*) FROM (
    (SELECT m, c, v, n, d FROM clean_calc
     EXCEPT SELECT reporting_month, product_category, item_sales_value, items_sold, distinct_orders FROM curated.finance_monthly_category_sales)
    UNION ALL
    (SELECT reporting_month, product_category, item_sales_value, items_sold, distinct_orders FROM curated.finance_monthly_category_sales
     EXCEPT SELECT m, c, v, n, d FROM clean_calc)
) x
"""
check("FIN-CUR", "every (month, category) row matches CLEAN recalculation (value, items, distinct orders)",
      scalar(diff_sql) == 0, "differing rows = 0")

# 4) Distinct orders: per month the true distinct count is <= SUM over categories (multi-category orders)
cmp = query_df("""
    SELECT c.reporting_month,
           SUM(c.distinct_orders) AS sum_of_category_distinct_orders,
           (SELECT COUNT(DISTINCT s.order_id) FROM staging.finance_items_reportable s
             WHERE s.reporting_month = c.reporting_month) AS true_distinct_orders
    FROM curated.finance_monthly_category_sales c GROUP BY c.reporting_month ORDER BY 1
""")
check("FIN-CUR", "sum of per-category distinct_orders >= true distinct orders in every month (so it must not be summed)",
      bool((cmp.sum_of_category_distinct_orders >= cmp.true_distinct_orders).all()),
      f"months checked={len(cmp)}")
cmp.tail(6)

PASS | FIN-CUR | curated row count = distinct (month, category) pairs | rows=1,282
PASS | FIN-CUR | no duplicate (reporting_month, product_category) | 
PASS | FIN-CUR | total item_sales_value = CLEAN total (same rules) | clean=13,494,400.74 curated=13,494,400.74


PASS | FIN-CUR | total items_sold = CLEAN eligible item count | 


PASS | FIN-CUR | every (month, category) row matches CLEAN recalculation (value, items, distinct orders) | differing rows = 0


PASS | FIN-CUR | sum of per-category distinct_orders >= true distinct orders in every month (so it must not be summed) | months checked=24


,reporting_month,sum_of_category_distinct_orders,true_distinct_orders
18,2018-04-01,6987,6919
19,2018-05-01,6904,6833
20,2018-06-01,6208,6145
21,2018-07-01,6286,6233
22,2018-08-01,6494,6421
23,2018-09-01,1,1


### Final business query - Finance
Reads **only** from `curated`. No joins, no business logic.

In [ ]:
# Monthly total item sales (all categories)
query_df("""
    SELECT reporting_month,
           SUM(item_sales_value) AS item_sales_value,
           SUM(items_sold)       AS items_sold
    FROM curated.finance_monthly_category_sales
    GROUP BY reporting_month
    ORDER BY reporting_month
""")

,reporting_month,item_sales_value,items_sold
0,2016-09-01,207.86,5
1,2016-10-01,44507.30,342
2,2016-12-01,10.90,1
3,2017-01-01,120098.27,953
4,2017-02-01,244959.35,1936
5,2017-03-01,368341.32,2975
6,2017-04-01,353842.98,2660
7,2017-05-01,503159.19,4106
8,2017-06-01,429916.61,3571
9,2017-07-01,492287.30,4493


In [ ]:
# Top 10 categories in one month
query_df("""
    SELECT product_category, item_sales_value, items_sold, distinct_orders
    FROM curated.finance_monthly_category_sales
    WHERE reporting_month = DATE '2018-07-01'
    ORDER BY item_sales_value DESC
    LIMIT 10
""")

,product_category,item_sales_value,items_sold,distinct_orders
0,beleza_saude,105517.68,782,711
1,relogios_presentes,96212.59,528,514
2,utilidades_domesticas,63146.37,571,469
3,esporte_lazer,55172.78,466,416
4,cama_mesa_banho,54904.92,621,515
5,moveis_decoracao,45659.39,438,345
6,automotivo,43564.79,342,325
7,informatica_acessorios,41581.48,424,386
8,telefonia,26539.51,224,205
9,ferramentas_jardim,26173.78,179,154


---
# 06. OPERATIONS PIPELINE

**Requirement -> Grain -> CLEAN sources -> STAGING 1 -> STAGING 2 -> CURATED -> Business query**

Finance thinks *events -> measures -> aggregation*. Operations is different: *orders -> dates -> delivery status -> business classification -> aggregation*. So the staging design is different too.

## Step 1 - Understand
Operations wants to know whether delivery performance is getting better or worse over time and **which states** have late deliveries. Definitions are in section 01.

## Step 2 - Grain
* The metric is about **orders**, so the working grain is **one row per order**. If I started from `order_items` an order with 5 items would be counted 5 times.
* Final grain: **one customer state in one purchase month**.

## Step 3 & 4 - Sources and relationships
* `clean.orders` holds the status, purchase date, **estimated** delivery date (the promise) and **actual** delivery date.
* `clean.customers` holds the state/city. Relationship: `orders.customer_id -> customers.customer_id` (many orders to one customer record; validated with zero orphans).
* `order_items` and `reviews` are **not** needed.

## Check the data before defining "late" (evidence)

In [ ]:
query_df("""
    SELECT order_status,
           COUNT(*)                                         AS orders,
           COUNT(order_delivered_customer_date)             AS has_delivery_date,
           COUNT(*) - COUNT(order_delivered_customer_date)  AS missing_delivery_date,
           COUNT(*) FILTER (WHERE CAST(order_delivered_customer_date AS date) > CAST(order_estimated_delivery_date AS date)) AS after_estimate
    FROM clean.orders GROUP BY order_status ORDER BY orders DESC
""")

,order_status,orders,has_delivery_date,missing_delivery_date,after_estimate
0,delivered,96478,96470,8,6534
1,shipped,1107,0,1107,0
2,canceled,625,6,619,1
3,unavailable,609,0,609,0
4,invoiced,314,0,314,0
5,processing,301,0,301,0
6,created,5,0,5,0
7,approved,2,0,2,0


In [ ]:
# Boundary evidence: how many orders arrive the SAME DAY as the estimate but later than midnight?
# (This is why I compare dates, not timestamps.)
print("Delivered on the estimated day (after 00:00):", scalar("""
    SELECT COUNT(*) FROM clean.orders
    WHERE order_status = 'delivered'
      AND CAST(order_delivered_customer_date AS date) = CAST(order_estimated_delivery_date AS date)
"""))
print("Missing estimated delivery date:", null_count("clean.orders", "order_estimated_delivery_date"))
print("Delivered orders with NO delivery date:", scalar("SELECT COUNT(*) FROM clean.orders WHERE order_status = 'delivered' AND order_delivered_customer_date IS NULL"))
print("Canceled orders that DO have a delivery date:", scalar("SELECT COUNT(*) FROM clean.orders WHERE order_status = 'canceled' AND order_delivered_customer_date IS NOT NULL"))

Delivered on the estimated day (after 00:00): 1292
Missing estimated delivery date: 0
Delivered orders with NO delivery date: 8
Canceled orders that DO have a delivery date: 6


**What the evidence tells me**
* Only `delivered` orders (almost all of them) have a delivery date; a few `delivered` orders have none -> `unknown`.
* A handful of `canceled` orders have a delivery date - status wins, they stay `not_fulfilled`.
* Many orders arrive on the estimated day, so the boundary rule (same day = on time) really matters.

## STAGING 1 - `staging.ops_orders_base`
* **One row =** one order.
* **Why it exists:** collects exactly the evidence needed to judge delivery (identity, status, purchase date, promised date, actual date, customer location) - and nothing is classified yet, so each order can be inspected before any rule is applied.
* **Prediction:** same row count as `clean.orders` (99,441). `orders -> customers` is many-to-one and every order has a customer, so an `INNER JOIN` neither multiplies nor drops rows.

In [ ]:
sql_ops_stg1 = """
DROP TABLE IF EXISTS staging.ops_orders_base;

CREATE TABLE staging.ops_orders_base AS
SELECT
    o.order_id,
    o.customer_id,
    c.customer_state,
    c.customer_city,
    o.order_status,
    o.order_purchase_timestamp,
    o.order_estimated_delivery_date,
    o.order_delivered_customer_date
FROM clean.orders o
JOIN clean.customers c ON c.customer_id = o.customer_id;

ALTER TABLE staging.ops_orders_base ADD PRIMARY KEY (order_id)
"""
run_sql(sql_ops_stg1)

In [ ]:
expected = scalar("SELECT COUNT(*) FROM clean.orders")
actual = scalar("SELECT COUNT(*) FROM staging.ops_orders_base")
check("OPS-STG1", "row count = clean.orders (join neither multiplied nor dropped orders)", expected == actual,
      f"expected={expected:,} actual={actual:,}")
check("OPS-STG1", "declared grain unique (order_id)", dup_count("staging.ops_orders_base", ["order_id"]) == 0)
for col in ["order_id", "customer_state", "order_status", "order_purchase_timestamp", "order_estimated_delivery_date"]:
    check("OPS-STG1", f"no NULLs in {col}", null_count("staging.ops_orders_base", col) == 0)
print("Info: NULL delivery dates (expected, undelivered/cancelled orders):",
      f"{null_count('staging.ops_orders_base', 'order_delivered_customer_date'):,}")
query_df("SELECT * FROM staging.ops_orders_base ORDER BY order_id LIMIT 5")

PASS | OPS-STG1 | row count = clean.orders (join neither multiplied nor dropped orders) | expected=99,441 actual=99,441
PASS | OPS-STG1 | declared grain unique (order_id) | 
PASS | OPS-STG1 | no NULLs in order_id | 
PASS | OPS-STG1 | no NULLs in customer_state | 
PASS | OPS-STG1 | no NULLs in order_status | 
PASS | OPS-STG1 | no NULLs in order_purchase_timestamp | 
PASS | OPS-STG1 | no NULLs in order_estimated_delivery_date | 
Info: NULL delivery dates (expected, undelivered/cancelled orders): 2,965


,order_id,customer_id,customer_state,customer_city,order_status,order_purchase_timestamp,order_estimated_delivery_date,order_delivered_customer_date
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,RJ,campos dos goytacazes,delivered,2017-09-13 08:59:02,2017-09-29,2017-09-20 23:43:48
1,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,SP,santa fe do sul,delivered,2017-04-26 10:53:06,2017-05-15,2017-05-12 16:04:24
2,000229ec398224ef6ca0657da4fc703e,6489ae5e4333f3693df5ad4372dab6d3,MG,para de minas,delivered,2018-01-14 14:33:31,2018-02-05,2018-01-22 13:19:16
3,00024acbcdf0a6daa1e931b038114c75,d4eb9395c8c0431ee92fce09860c5a06,SP,atibaia,delivered,2018-08-08 10:00:35,2018-08-20,2018-08-14 13:32:39
4,00042b26cf59d7ce69dfabb4e55b4fd9,58dbd0b2d70206bf40e62cd34e84d795,SP,varzea paulista,delivered,2017-02-04 13:57:51,2017-03-17,2017-03-01 16:42:31


## STAGING 2 - `staging.ops_orders_classified`
* **One row =** one order, now with a delivery classification.
* **Why it exists:** the business rule "what counts as late" lives in exactly one place. After this table, the final aggregation is just counting.
* **What changed from staging 1:** added `reporting_month`, `days_vs_estimate` (negative = early, positive = late) and `delivery_classification`.

| classification | evidence rule |
|---|---|
| `not_fulfilled` | status is `canceled` or `unavailable` |
| `unknown` | status `delivered` but delivery date is missing |
| `late` | status `delivered` and delivery **date** is after estimated **date** |
| `on_time` | status `delivered` and delivery date <= estimated date (same day = on time) |
| `not_delivered` | any other status (created, approved, invoiced, processing, shipped) - still in progress |

The `CASE` is evaluated top-down, so cancelled orders with a delivery date are still `not_fulfilled`.

* **Prediction:** same row count as staging 1 (a classification adds a column, it does not add or remove rows), and every order gets exactly one class (no NULL class).

In [ ]:
sql_ops_stg2 = """
DROP TABLE IF EXISTS staging.ops_orders_classified;

CREATE TABLE staging.ops_orders_classified AS
SELECT
    b.order_id,
    b.customer_state,
    b.customer_city,
    b.order_status,
    CAST(date_trunc('month', b.order_purchase_timestamp) AS date) AS reporting_month,
    b.order_estimated_delivery_date,
    b.order_delivered_customer_date,
    CASE
        WHEN b.order_delivered_customer_date IS NOT NULL
         AND b.order_status = 'delivered'
        THEN CAST(b.order_delivered_customer_date AS date) - CAST(b.order_estimated_delivery_date AS date)
    END AS days_vs_estimate,
    CASE
        WHEN b.order_status IN ('canceled', 'unavailable')                                   THEN 'not_fulfilled'
        WHEN b.order_status = 'delivered' AND b.order_delivered_customer_date IS NULL         THEN 'unknown'
        WHEN b.order_status = 'delivered'
         AND CAST(b.order_delivered_customer_date AS date) > CAST(b.order_estimated_delivery_date AS date) THEN 'late'
        WHEN b.order_status = 'delivered'                                                    THEN 'on_time'
        ELSE 'not_delivered'
    END AS delivery_classification
FROM staging.ops_orders_base b;

ALTER TABLE staging.ops_orders_classified ADD PRIMARY KEY (order_id)
"""
run_sql(sql_ops_stg2)

In [ ]:
check("OPS-STG2", "row count = staging 1 (classification did not change grain)",
      scalar("SELECT COUNT(*) FROM staging.ops_orders_classified") == scalar("SELECT COUNT(*) FROM staging.ops_orders_base"))
check("OPS-STG2", "declared grain unique (order_id)", dup_count("staging.ops_orders_classified", ["order_id"]) == 0)
check("OPS-STG2", "every order has exactly one classification (no NULL)", null_count("staging.ops_orders_classified", "delivery_classification") == 0)
check("OPS-STG2", "every order has a reporting_month", null_count("staging.ops_orders_classified", "reporting_month") == 0)

dist = query_df("""
    SELECT delivery_classification, COUNT(*) AS orders,
           MIN(days_vs_estimate) AS min_days_vs_estimate, MAX(days_vs_estimate) AS max_days_vs_estimate
    FROM staging.ops_orders_classified GROUP BY 1 ORDER BY orders DESC
""")
display(dist)

# Boundary + sign checks: late <=> days_vs_estimate > 0 ; on_time <=> days_vs_estimate <= 0
bad_late = scalar("SELECT COUNT(*) FROM staging.ops_orders_classified WHERE delivery_classification = 'late' AND days_vs_estimate <= 0")
bad_ontime = scalar("SELECT COUNT(*) FROM staging.ops_orders_classified WHERE delivery_classification = 'on_time' AND days_vs_estimate > 0")
check("OPS-STG2", "late orders always have days_vs_estimate > 0; on_time never > 0", bad_late == 0 and bad_ontime == 0)
check("OPS-STG2", "same-day deliveries (days_vs_estimate = 0) are on_time",
      scalar("SELECT COUNT(*) FROM staging.ops_orders_classified WHERE days_vs_estimate = 0 AND delivery_classification <> 'on_time'") == 0)
check("OPS-STG2", "cancelled/unavailable orders are never late",
      scalar("SELECT COUNT(*) FROM staging.ops_orders_classified WHERE order_status IN ('canceled','unavailable') AND delivery_classification <> 'not_fulfilled'") == 0)
query_df("SELECT * FROM staging.ops_orders_classified WHERE delivery_classification = 'late' ORDER BY order_id LIMIT 5")

PASS | OPS-STG2 | row count = staging 1 (classification did not change grain) | 
PASS | OPS-STG2 | declared grain unique (order_id) | 
PASS | OPS-STG2 | every order has exactly one classification (no NULL) | 
PASS | OPS-STG2 | every order has a reporting_month | 


,delivery_classification,orders,min_days_vs_estimate,max_days_vs_estimate
0,on_time,89936,-147.0,0.0
1,late,6534,1.0,188.0
2,not_delivered,1729,NaN,NaN
3,not_fulfilled,1234,NaN,NaN
4,unknown,8,NaN,NaN


PASS | OPS-STG2 | late orders always have days_vs_estimate > 0; on_time never > 0 | 
PASS | OPS-STG2 | same-day deliveries (days_vs_estimate = 0) are on_time | 
PASS | OPS-STG2 | cancelled/unavailable orders are never late | 


,order_id,customer_state,customer_city,order_status,reporting_month,order_estimated_delivery_date,order_delivered_customer_date,days_vs_estimate,delivery_classification
0,000e906b789b55f64edcb1f84030f90d,SP,santa cruz do rio pardo,delivered,2017-11-01,2017-12-07,2017-12-09 17:27:23,2,late
1,0017afd5076e074a48f1f1a4c7bac9c5,AL,campo alegre,delivered,2017-04-01,2017-05-19,2017-05-23 08:32:07,4,late
2,001c85b5f68d2be0cb0797afc9e8ce9a,SP,sao paulo,delivered,2017-11-01,2017-12-14,2017-12-22 18:37:40,8,late
3,001d8f0e34a38c37f7dba2a37d4eba8b,SP,sao paulo,delivered,2017-05-01,2017-05-24,2017-05-26 13:14:50,2,late
4,0030d783f979fbc5981e75613b057344,RJ,rio de janeiro,delivered,2017-11-01,2017-12-15,2018-01-06 00:25:43,22,late


## CURATED - `curated.operations_monthly_state_delivery`
* **Purpose:** let Operations see delivery performance by month and state without re-deriving what "late" means.
* **One row =** one customer state in one purchase month.
* **Columns / metrics**
  * `reporting_month` - first day of the purchase month
  * `customer_state` - 2-letter state of the customer
  * `total_orders` - all orders purchased in that month from that state (any status)
  * `delivered_orders` - orders classified `late` or `on_time` (trustworthy delivery outcome)
  * `late_orders` - orders classified `late`
  * `late_delivery_rate` - `late_orders / delivered_orders` (NULL when no delivered orders)
* **Prediction:** `SUM(total_orders)` = 99,441 = all orders; `late_orders <= delivered_orders <= total_orders` on every row.

In [ ]:
sql_ops_curated = """
DROP TABLE IF EXISTS curated.operations_monthly_state_delivery;

CREATE TABLE curated.operations_monthly_state_delivery (
    reporting_month     DATE         NOT NULL,
    customer_state      CHAR(2)      NOT NULL,
    total_orders        INTEGER      NOT NULL,
    delivered_orders    INTEGER      NOT NULL,
    late_orders         INTEGER      NOT NULL,
    late_delivery_rate  NUMERIC(6,4),
    PRIMARY KEY (reporting_month, customer_state)
);

INSERT INTO curated.operations_monthly_state_delivery
WITH counts AS (
    SELECT
        reporting_month,
        customer_state,
        COUNT(*)                                                                   AS total_orders,
        COUNT(*) FILTER (WHERE delivery_classification IN ('late', 'on_time'))     AS delivered_orders,
        COUNT(*) FILTER (WHERE delivery_classification = 'late')                   AS late_orders
    FROM staging.ops_orders_classified
    GROUP BY reporting_month, customer_state
)
SELECT
    reporting_month, customer_state, total_orders, delivered_orders, late_orders,
    ROUND(CAST(late_orders AS numeric) / NULLIF(delivered_orders, 0), 4)
FROM counts;

COMMENT ON TABLE curated.operations_monthly_state_delivery IS
  'Operations product. Grain: one customer state per purchase month. delivered = status delivered with a delivery date, late = delivery date > estimated date (date compare, same day = on time), late_delivery_rate = late_orders / delivered_orders.'
"""
run_sql(sql_ops_curated)

### Reconcile - CURATED vs CLEAN (independent recalculation, no staging)

In [ ]:
# 1) Grain
check("OPS-CUR", "no duplicate (reporting_month, customer_state)",
      dup_count("curated.operations_monthly_state_delivery", ["reporting_month", "customer_state"]) == 0)
check("OPS-CUR", "row count = distinct (month, state) pairs in staging 2",
      scalar("SELECT COUNT(*) FROM curated.operations_monthly_state_delivery") ==
      scalar("SELECT COUNT(*) FROM (SELECT DISTINCT reporting_month, customer_state FROM staging.ops_orders_classified) d"))

# 2) Totals vs CLEAN
check("OPS-CUR", "SUM(total_orders) = clean.orders row count (no order lost or double counted)",
      scalar("SELECT SUM(total_orders) FROM curated.operations_monthly_state_delivery") == scalar("SELECT COUNT(*) FROM clean.orders"))
clean_delivered = scalar("SELECT COUNT(*) FROM clean.orders WHERE order_status = 'delivered' AND order_delivered_customer_date IS NOT NULL")
clean_late = scalar("""
    SELECT COUNT(*) FROM clean.orders
    WHERE order_status = 'delivered' AND order_delivered_customer_date IS NOT NULL
      AND CAST(order_delivered_customer_date AS date) > CAST(order_estimated_delivery_date AS date)
""")
check("OPS-CUR", "SUM(delivered_orders) = CLEAN delivered orders", scalar("SELECT SUM(delivered_orders) FROM curated.operations_monthly_state_delivery") == clean_delivered,
      f"delivered={clean_delivered:,}")
check("OPS-CUR", "SUM(late_orders) = CLEAN late orders", scalar("SELECT SUM(late_orders) FROM curated.operations_monthly_state_delivery") == clean_late,
      f"late={clean_late:,}")

# 3) Logical bounds
check("OPS-CUR", "late <= delivered <= total on every row",
      scalar("SELECT COUNT(*) FROM curated.operations_monthly_state_delivery WHERE NOT (late_orders <= delivered_orders AND delivered_orders <= total_orders)") == 0)
check("OPS-CUR", "late_delivery_rate between 0 and 1 (or NULL when no delivered orders)",
      scalar("SELECT COUNT(*) FROM curated.operations_monthly_state_delivery WHERE late_delivery_rate < 0 OR late_delivery_rate > 1") == 0
      and scalar("SELECT COUNT(*) FROM curated.operations_monthly_state_delivery WHERE (delivered_orders = 0) <> (late_delivery_rate IS NULL)") == 0)

# 4) Month x state: differing rows in either direction must be 0
diff_sql = """
WITH clean_calc AS (
    SELECT CAST(date_trunc('month', o.order_purchase_timestamp) AS date) AS m,
           c.customer_state AS s,
           COUNT(*) AS t,
           SUM(CASE WHEN o.order_status = 'delivered' AND o.order_delivered_customer_date IS NOT NULL THEN 1 ELSE 0 END) AS d,
           SUM(CASE WHEN o.order_status = 'delivered' AND o.order_delivered_customer_date IS NOT NULL
                     AND CAST(o.order_delivered_customer_date AS date) > CAST(o.order_estimated_delivery_date AS date) THEN 1 ELSE 0 END) AS l
    FROM clean.orders o JOIN clean.customers c ON c.customer_id = o.customer_id
    GROUP BY 1, 2
)
SELECT COUNT(*) FROM (
    (SELECT m, s, t, d, l FROM clean_calc
     EXCEPT SELECT reporting_month, customer_state, total_orders, delivered_orders, late_orders FROM curated.operations_monthly_state_delivery)
    UNION ALL
    (SELECT reporting_month, customer_state, total_orders, delivered_orders, late_orders FROM curated.operations_monthly_state_delivery
     EXCEPT SELECT m, s, t, d, l FROM clean_calc)
) x
"""
check("OPS-CUR", "every (month, state) row matches CLEAN recalculation", scalar(diff_sql) == 0, "differing rows = 0")

PASS | OPS-CUR | no duplicate (reporting_month, customer_state) | 
PASS | OPS-CUR | row count = distinct (month, state) pairs in staging 2 | 
PASS | OPS-CUR | SUM(total_orders) = clean.orders row count (no order lost or double counted) | 
PASS | OPS-CUR | SUM(delivered_orders) = CLEAN delivered orders | delivered=96,470
PASS | OPS-CUR | SUM(late_orders) = CLEAN late orders | late=6,534
PASS | OPS-CUR | late <= delivered <= total on every row | 
PASS | OPS-CUR | late_delivery_rate between 0 and 1 (or NULL when no delivered orders) | 


PASS | OPS-CUR | every (month, state) row matches CLEAN recalculation | differing rows = 0


### Final business query - Operations
Reads **only** from `curated`. It does not need to know what "late" means.

In [ ]:
# How is delivery performance changing over time? (national monthly late rate)
query_df("""
    SELECT reporting_month,
           SUM(total_orders)     AS total_orders,
           SUM(delivered_orders) AS delivered_orders,
           SUM(late_orders)      AS late_orders,
           ROUND(CAST(SUM(late_orders) AS numeric) / NULLIF(SUM(delivered_orders), 0), 4) AS late_delivery_rate
    FROM curated.operations_monthly_state_delivery
    GROUP BY reporting_month
    ORDER BY reporting_month
""")

,reporting_month,total_orders,delivered_orders,late_orders,late_delivery_rate
0,2016-09-01,4,1,1,1.0000
1,2016-10-01,324,265,2,0.0075
2,2016-12-01,1,1,0,0.0000
3,2017-01-01,800,750,22,0.0293
4,2017-02-01,1780,1653,49,0.0296
5,2017-03-01,2682,2546,116,0.0456
6,2017-04-01,2404,2303,151,0.0656
7,2017-05-01,3700,3545,106,0.0299
8,2017-06-01,3245,3135,95,0.0303
9,2017-07-01,4026,3872,108,0.0279


In [ ]:
# Where are late deliveries occurring? (states, 2018, enough volume to be meaningful)
query_df("""
    SELECT customer_state,
           SUM(delivered_orders) AS delivered_orders,
           SUM(late_orders)      AS late_orders,
           ROUND(CAST(SUM(late_orders) AS numeric) / NULLIF(SUM(delivered_orders), 0), 4) AS late_delivery_rate
    FROM curated.operations_monthly_state_delivery
    WHERE reporting_month BETWEEN DATE '2018-01-01' AND DATE '2018-08-01'
    GROUP BY customer_state
    HAVING SUM(delivered_orders) >= 100
    ORDER BY late_delivery_rate DESC
    LIMIT 10
""")

,customer_state,delivered_orders,late_orders,late_delivery_rate
0,AL,198,44,0.2222
1,MA,346,69,0.1994
2,CE,633,121,0.1912
3,PI,258,43,0.1667
4,PA,453,75,0.1656
5,SE,146,24,0.1644
6,TO,144,22,0.1528
7,MS,411,59,0.1436
8,BA,1726,243,0.1408
9,ES,1047,144,0.1375


---
# 07. FINAL VALIDATION & REFLECTION

Every check recorded by `check(...)` during this run is summarised below and written to `docs/validation_evidence.md`.

In [ ]:
evidence = pd.DataFrame(CHECKS)
print("Checks run:", len(evidence), "| passed:", int((evidence.result == "PASS").sum()), "| failed:", int((evidence.result == "FAIL").sum()))
display(evidence.groupby("stage", sort=False).result.agg(checks="count", passed=lambda s: int((s == "PASS").sum())))

Checks run: 69 | passed: 69 | failed: 0


,checks,passed
stage,,
RAW,7,7
CLEAN,21,21
FIN-STG1,7,7
FIN-STG2,6,6
FIN-CUR,6,6
OPS-STG1,7,7
OPS-STG2,7,7
OPS-CUR,8,8


In [ ]:
# Row counts of every table in every layer (the full trail from RAW to CURATED)
inventory = query_df("""
    SELECT table_schema AS layer, table_name FROM information_schema.tables
    WHERE table_schema IN ('raw','clean','staging','curated') ORDER BY
      CASE table_schema WHEN 'raw' THEN 1 WHEN 'clean' THEN 2 WHEN 'staging' THEN 3 ELSE 4 END, table_name
""")
inventory["rows"] = [scalar(f"SELECT COUNT(*) FROM {r.layer}.{r.table_name}") for r in inventory.itertuples()]
inventory

,layer,table_name,rows
0,raw,customers,99441
1,raw,order_items,112650
2,raw,orders,99441
3,raw,products,32951
4,raw,reviews,99224
5,clean,customers,99441
6,clean,order_items,112650
7,clean,orders,99441
8,clean,products,32951
9,clean,reviews,99224


In [ ]:
# Write validation evidence document
lines = [
    "# Validation Evidence",
    "",
    "Generated by `notebooks/assignment_3_warehouse.ipynb` (section 07). Re-run the notebook to refresh.",
    "",
    f"**Checks run:** {len(evidence)}  |  **Passed:** {int((evidence.result == 'PASS').sum())}  |  **Failed:** {int((evidence.result == 'FAIL').sum())}",
    "",
    "## Checks",
    "",
    "| Stage | Check | Result | Evidence |",
    "|---|---|---|---|",
]
for r in evidence.itertuples():
    lines.append(f"| {r.stage} | {r.check} | {r.result} | {r.detail} |")
lines += ["", "## Table inventory (rows per table)", "", "| Layer | Table | Rows |", "|---|---|---|"]
for r in inventory.itertuples():
    lines.append(f"| {r.layer} | {r.table_name} | {r.rows:,} |")
(DOCS_DIR / "validation_evidence.md").write_text("\n".join(lines) + "\n", encoding="utf-8")
print("Wrote", DOCS_DIR / "validation_evidence.md")

Wrote /home/claude/work/olist-pandas-assignment/docs/validation_evidence.md


## Traceability: final number -> raw evidence

```
curated.finance_monthly_category_sales           curated.operations_monthly_state_delivery
        ^  (GROUP BY month, category)                    ^  (GROUP BY month, state)
staging.finance_items_reportable                 staging.ops_orders_classified
        ^  (status filter, month, category)              ^  (late / on_time / ... rules)
staging.finance_items_enriched                   staging.ops_orders_base
        ^  (items + orders + products)                   ^  (orders + customers)
                         clean.* (typed, keyed entities)
                                   ^  (types restored with Pandas)
                    data/processed/*.csv  (Assignment 1, validated)
                                   ^
        raw.* (TEXT, from data/raw/*.csv - the original evidence)
```

## What I learned
* **Grain first.** Writing "one row = ..." before each table made it obvious where rows could multiply (items vs orders) and where `COUNT(DISTINCT)` is needed.
* **Staging gives each rule a home.** Joining, business filtering and classification are separate, inspectable steps instead of one giant query.
* **Predict, then verify.** Each prediction was compared with the database, and independent recalculation from CLEAN is what turns "the query ran" into "the number is right".
* **Definitions are decisions.** Excluding freight, comparing dates instead of timestamps, treating `canceled` as not late and using `delivered_orders` as the rate denominator all change the numbers - so each one is written down.

## What I would improve
* Add a category-translation table so Finance sees English category names.
* Turn the notebook sections into a re-runnable script / scheduled pipeline (next stage - intentionally out of scope here).
* Add data-quality alerts (e.g. fail the run if orphans or duplicate grain appear) instead of checking by hand.
* Investigate the few `delivered` orders with no delivery date and the cancelled orders that have one with the source team.